src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs

```diff
@@ -3327,4 +3344,404 @@ internal void SetDerivedNames(string protocol, string serverName)
+                        break;
+                    default:
+                        // None of the supported protocols were found. This may be a IPv6 address
+                        _connectionProtocol = Protocol.None;
+                        break;
+                }
+            }
+        }
+
+        // LocalDbInstance name always starts with (localdb)
+        // possible scenarios:
+        // (localdb)\<instance name>
+        // or (localdb)\. which goes to default localdb
+        // or (localdb)\.\<sharedInstance name>
+        internal static string GetLocalDBInstance(string dataSource, out bool error)
+        {
+            string instanceName = null;
+
+            // ReadOnlySpan is not supported in netstandard 2.0, but installing System.Memory solves the issue
+            ReadOnlySpan<char> input = dataSource.AsSpan().TrimStart();
+            error = false;
+
+            // NetStandard 2.0 does not support passing a string to ReadOnlySpan<char>
+            if (input.StartsWith(LocalDbHost.AsSpan().Trim(), StringComparison.InvariantCultureIgnoreCase))
+            {
+                // When netcoreapp support for netcoreapp2.1 is dropped these slice calls could be converted to System.Range\System.Index
+                // Such ad input = input[1..];
+                input = input.Slice(LocalDbHost.Length);
+                if (!input.IsEmpty && input[0] == BackSlashCharacter)
+                {
+                    input = input.Slice(1);
+                }
+                if (!input.IsEmpty)
+                {
+                    instanceName = input.Trim().ToString();
+                }
+                else
+                {
+                    error = true;
+                }
+            }
+            return instanceName;
+        }
+
+        internal static DataSource ParseServerName(string dataSource)
+        {
+            DataSource details = new DataSource(dataSource);
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (details.InferNamedPipesInformation())
+            {
+                return details;
+            }
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (details.InferConnectionDetails())
+            {
+                return details;
+            }
+
+            return null;
+        }
+
+        private void InferLocalServerName()
+        {
+            // If Server name is empty or localhost, then use "localhost"
+            if (string.IsNullOrEmpty(ServerName) || IsLocalHost(ServerName))
+            {
+                ServerName = _connectionProtocol == Protocol.Admin ?
+                    Environment.MachineName : DefaultHostName;
+            }
+        }
+
+        private bool InferConnectionDetails()
+        {
+            string[] tokensByCommaAndSlash = _dataSourceAfterTrimmingProtocol.Split(BackSlashCharacter, CommaSeparator);
```

Comment:
Could this use span to avoid the string and array allocations? We had some interesting back and forth on using spans while parsing in https://github.com/dotnet/SqlClient/pull/1237 that might be useful to review.

Reply:
It's definitely worth exploring. Everything said above applies.